pkg/workceptor/kubernetes.go

```diff
@@ -224,6 +236,106 @@ func podRunningAndReady(kw KubeUnit) func(event watch.Event) (bool, error) {
 	return inner
 }
 
+// podInfrastructureFailure checks if the pod has failed to start due to infrastructure issues.
+func (ku KubeAPIWrapper) podInfrastructureSuccess(pod *corev1.Pod) (bool, string, error) {
+	if pod == nil {
+		return false, "pod is nil", fmt.Errorf("pod is nil")
+	}
+
+	switch pod.Status.Phase {
+	case corev1.PodFailed:
+		for _, cs := range pod.Status.ContainerStatuses {
+			if cs.State.Terminated != nil {
+				switch cs.State.Terminated.Reason {
+				case "OOMKilled", "Evicted":
+					return false, cs.State.Terminated.Reason, nil
+				}
+			}
+		}
+		return true, "", nil
+
+	case corev1.PodSucceeded:
+		return true, "", nil
+
+	case corev1.PodRunning, corev1.PodPending:
+		return true, fmt.Sprintf("pod ohase: %s", pod.Status.Phase), fmt.Errorf("pod not in terminal state")
+
+	default:
+		return false, fmt.Sprintf("unknown phase: %s", pod.Status.Phase), fmt.Errorf("invalid pod phase")
+	}
+}
+
+// PodApplicationSuccess checks if the pod has successfully completed its application logic.
+func (ku KubeAPIWrapper) podApplicationSuccess(pod *corev1.Pod) (bool, string, error) {
+	if pod == nil {
+		return false, "pod is nil", fmt.Errorf("pod is nil")
+	}
+
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.State.Terminated == nil {
+			return true, "container has not terminated", nil
+		}
+		if cs.State.Terminated.ExitCode != 0 {
+			return false, fmt.Sprintf("container exited with code %d: %s", cs.State.Terminated.ExitCode, cs.State.Terminated.Reason), nil
+		}
+		if cs.State.Terminated.Reason == "Error" {
+			return false, "container terminated with 'Error'", nil
+		}
```

Comment:
```suggestion
		if cs.State.Terminated.ExitCode != 0 {
			return false, fmt.Sprintf("container exited with code %d: %s", cs.State.Terminated.ExitCode, cs.State.Terminated.Reason), nil
		}
```

`ExitCode` and `Reason` are returned at the same time `cs.State.Terminated.Reason == "Error"` is included in the `cs.State.Terminated.ExitCode != 0` statement so lines 281-283 is unreachable code.